File: src/main/java/me/muksc/tacztweaks/network/message/ClientMessagePlayerUnload.java
package me.muksc.tacztweaks.network.message;

import com.tacz.guns.api.item.IGun;
import com.tacz.guns.entity.shooter.ShooterDataHolder;
import me.muksc.tacztweaks.Config;
import me.muksc.tacztweaks.ShooterDataHolderProvider;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.network.NetworkEvent;

import java.util.function.Supplier;

public class ClientMessagePlayerUnload {
    public ClientMessagePlayerUnload() { /* Nothing */ }

    public static void encode(ClientMessagePlayerUnload message, FriendlyByteBuf buf) { /* Nothing */ }

    public static ClientMessagePlayerUnload decode(FriendlyByteBuf buf) {
        return new ClientMessagePlayerUnload();
    }

    public static void handle(ClientMessagePlayerUnload message, Supplier<NetworkEvent.Context> contextSupplier) {
        NetworkEvent.Context context = contextSupplier.get();
        if (!context.getDirection().getReceptionSide().isServer() || !Config.Gun.INSTANCE.allowUnload()) {
            context.setPacketHandled(true);
            return;
        }

        context.enqueueWork(() -> {
            ServerPlayer player = context.getSender();
            if (player == null || player.isCreative()) return;
            ShooterDataHolder data = ((ShooterDataHolderProvider) player).tacztweaks$getShooterDataHolder();
            if (data.currentGunItem == null) return;
            ItemStack gunStack = data.currentGunItem.get();
            IGun gun = IGun.getIGunOrNull(gunStack);
            if (gun == null) return;
            gun.dropAllAmmo(player, gunStack);
        });
        context.setPacketHandled(true);
    }
}

File: src/main/kotlin/me/muksc/tacztweaks/network/message/ClientMessagePlayerUnload.kt
package me.muksc.tacztweaks.network.message

import com.tacz.guns.api.entity.IGunOperator
import com.tacz.guns.api.item.IGun
import me.muksc.tacztweaks.TaCZTweaks
import me.muksc.tacztweaks.config.Config
import me.muksc.tacztweaks.core.network.CustomPacketPayload
import me.muksc.tacztweaks.core.network.CustomPacketPayloadType
import me.muksc.tacztweaks.core.codec.StreamCodec
import me.muksc.tacztweaks.core.extension.hasCreativeAmmoBox
import me.muksc.tacztweaks.mixininterface.feature.keyactions.unload.UnloadableGun
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.server.MinecraftServer
import net.minecraft.server.level.ServerPlayer

object ClientMessagePlayerUnload : CustomPacketPayload<ClientMessagePlayerUnload> {
    val TYPE = CustomPacketPayloadType<ClientMessagePlayerUnload>(
        TaCZTweaks.id("client_player_unload")
    )
    val STREAM_CODEC: StreamCodec<FriendlyByteBuf, ClientMessagePlayerUnload> =
        StreamCodec.unit(ClientMessagePlayerUnload)

    fun handle(packet: ClientMessagePlayerUnload, server: MinecraftServer, player: ServerPlayer?) = server.execute {
        if (!Config.KeyActions.Unload.enabled() || player == null) return@execute
        val operator = IGunOperator.fromLivingEntity(player)
        val gunStack = operator.dataHolder.currentGunItem?.get() ?: return@execute
        if (player.inventory.hasCreativeAmmoBox(gunStack)) return@execute
        val gun = IGun.getIGunOrNull(gunStack) ?: return@execute
        UnloadableGun.of(gun).`tacztweaks$unload`(player, gunStack)
    }

    override fun self(): ClientMessagePlayerUnload = this

    override fun type(): CustomPacketPayloadType<ClientMessagePlayerUnload> = TYPE

    override fun codec(): StreamCodec<FriendlyByteBuf, ClientMessagePlayerUnload> = STREAM_CODEC
}